File: LeetCode/Easy/202_Happy_Number/solution.java
/*
LeetCode 202
Happy Number

Difficulty: Easy

Approach:
Use Floyd's cycle detection. Compute the sum of squares of digits
repeatedly with a slow and fast pointer. If fast reaches 1, the
number is happy. If slow meets fast, a cycle exists and it's not happy.

Time Complexity: O(log n)
Space Complexity: O(1)
*/

class Solution {
    private int getNext(int n) {
        int total = 0;
        while (n > 0) {
            int digit = n % 10;
            total += digit * digit;
            n /= 10;
        }
        return total;
    }

    public boolean isHappy(int n) {
        int slow = n;
        int fast = getNext(n);
        while (fast != 1 && slow != fast) {
            slow = getNext(slow);
            fast = getNext(getNext(fast));
        }
        return fast == 1;
    }
}

File: LeetCode/Easy/202_Happy_Number/solution.py
"""
LeetCode 202
Happy Number

Difficulty: Easy

Approach:
Use Floyd's cycle detection. Compute the sum of squares of digits
repeatedly with a slow and fast pointer. If fast reaches 1, the
number is happy. If slow meets fast, a cycle exists and it's not happy.

Time Complexity: O(log n)
Space Complexity: O(1)
"""


class Solution:
    def isHappy(self, n: int) -> bool:
        def get_next(number):
            total = 0
            while number > 0:
                digit = number % 10
                total += digit * digit
                number //= 10
            return total

        slow = n
        fast = get_next(n)
        while fast != 1 and slow != fast:
            slow = get_next(slow)
            fast = get_next(get_next(fast))
        return fast == 1
